packages/react-native-branch/src/withBranchAndroid.ts

```diff
@@ -0,0 +1,339 @@
+  assert(packageName, "android.package must be defined");
+
+  const mainApplicationPath = path.join(
+    config.modRequest.platformProjectRoot,
+    "app",
+    "src",
+    "main",
+    "java",
+    ...packageName.split("."),
+    "MainApplication.java"
+  );
+
+  try {
+    const mainApplication = action(await readFileAsync(mainApplicationPath));
+    return await saveFileAsync(mainApplicationPath, mainApplication);
+  } catch (e) {
+    WarningAggregator.addWarningAndroid(
+      "@config-plugins/react-native-branch",
+      `Couldn't modify MainApplication.java - ${e}.`
+    );
+  }
+}
+
+export async function editProguardRules(
+  config: ExportedConfigWithProps,
+  action: (mainApplication: string) => string
+) {
+  const proguardRulesPath = path.join(
+    config.modRequest.platformProjectRoot,
+    "app",
+    "proguard-rules.pro"
+  );
+  try {
+    const proguardRules = action(await readFileAsync(proguardRulesPath));
+    return await saveFileAsync(proguardRulesPath, proguardRules);
+  } catch (e) {
+    WarningAggregator.addWarningAndroid(
+      "@config-plugins/react-native-branch",
+      `Couldn't modify proguard-rules.pro - ${e}.`
+    );
+  }
+}
+
+export function addBranchMainApplicationImport(
+  src: string,
+  packageId: string
+): MergeResults {
+  const newSrc = ["import io.branch.rnbranch.RNBranchModule;"];
+
+  return mergeContents({
+    tag: "react-native-branch-import",
+    src,
+    newSrc: newSrc.join("\n"),
+    anchor: `package ${packageId};`,
+    offset: 1,
+    comment: "//",
+  });
+}
+
+export function addBranchGetAutoInstance(src: string): MergeResults {
+  const newSrc = ["    RNBranchModule.getAutoInstance(this);"];
+
+  return mergeContents({
+    tag: "react-native-branch-auto-instance",
+    src,
+    newSrc: newSrc.join("\n"),
+    anchor: /super\.onCreate\(\);/,
+    offset: 1,
+    comment: "//",
+  });
+}
+
+export function addBranchMainActivityImport(
+  src: string,
+  packageId: string
+): MergeResults {
+  const newSrc = [
+    "import android.content.Intent;",
+    "import io.branch.rnbranch.*;",
+  ];
+
+  return mergeContents({
+    tag: "react-native-branch-import",
+    src,
+    newSrc: newSrc.join("\n"),
+    anchor: `package ${packageId};`,
+    offset: 1,
+    comment: "//",
+  });
+}
+
+export function addBranchInitSession(src: string): MergeResults {
+  const tag = "react-native-branch-init-session";
+
+  try {
+    const newSrc = [
+      "    RNBranchModule.initSession(getIntent().getData(), this);",
+    ];
+
+    return mergeContents({
+      tag,
+      src,
+      newSrc: newSrc.join("\n"),
+      anchor: /super\.onStart\(\);/,
+      offset: 1,
+      comment: "//",
+    });
+  } catch (err) {
+    if ((err as any).code !== "ERR_NO_MATCH") {
+      throw err;
+    }
+  }
+
+  const newSrc = [
+    "  @Override",
+    "  protected void onStart() {",
+    "    super.onStart();",
+    "    RNBranchModule.initSession(getIntent().getData(), this);",
```

Comment:
I'm not certain but I think we also plan to add/have a system for delegating these methods into the module as well. CC @tsapeta @Kudo 

Ideally we'd only use the config plugin for modifying the static files.

Reply:
as pointed out by @EvanBacon. we are planning an extensible system for making modules setup easier. we addressed some internal modules at first stage and we would like to hear more feedbacks from the communities if the design fit the use cases. let me summary in the android part:

- `MainApplication.onCreate`: this is feasible in sdk44 by the [`ApplicationLifecycleListener.onCreate`](https://github.com/expo/expo/blob/e46f2255d115c584af1017c74548a9afd93faba1/packages/expo-modules-core/android/src/main/java/expo/modules/core/interfaces/ApplicationLifecycleListener.java#L7)
- `MainActivity.onStart`: we didn't hook at `onStart` as ReactActivity now, is it possible to do it in `onCreate`? in this case we have [`ReactActivityLifecycleListener.onCreate`](https://github.com/expo/expo/blob/e46f2255d115c584af1017c74548a9afd93faba1/packages/expo-modules-core/android/src/main/java/expo/modules/core/interfaces/ReactActivityLifecycleListener.java#L7)
- `MainActivity.onNewIntent`: we didn't hook at this point now, but we can do it also in the `ReactActivityLifecycleListener`
- `AndroidManifest.xml`: the pr did really great for the `withAndroidManifest` static mod.
- `proguard-rules.pro`: that's not necessary to modify application proguard rules. a library can also promote proguard rules hint. learn more from [the official document](https://developer.android.com/studio/projects/android-library) and search the `consumerProguardFiles` keyword.

Reply:
Hi, @Kudo. In the case rn-branch. Where I can put those native codes? I can't modify the package react-native-branch. Still you continue using expo-branch to contain those codes?

Reply:
we may put them in `@config-plugins/react-native-branch` package temporarily and ideally we could create a pr for `react-native-branch` in the future.

Reply:
Hi @Kudo, please guide me! Our apps blocking by `expo-branch` serval months before I can have effort to do this. So, I want to do this fast as possible to upgrade to SDKv43 (or 44. 😄 )

Reply:
> AndroidManifest.xml: the pr did really great for the withAndroidManifest static mod

I cloned from `expo-branch`, ;)

Reply:
we are working on the document of expo modules system. after the document landed, there would be a guide there!

Reply:
Can you give some basic idea/steps need to be do? I will reference from another plugin to complete it.

Reply:
> MainActivity.onNewIntent: we didn't hook at this point now, but we can do it also in the ReactActivityLifecycleListener

Okay, I have been update to new expo modules system. This is the last missing piece. Hi @Kudo, When can I use it?

On iOS, Should we upgrade to new expo modules system too?

Reply:
awesome 🔥 `onNewIntent` will be available from sdk 44 with https://github.com/expo/expo/pull/15550.

for ios, your case are mostly to modify AppDelegate. the new AppDelegate Subscriber system would be easier for you. you can preview the document from @tsapeta : https://github.com/expo/expo/blob/%40tsapeta/creating-expo-modules/docs/pages/modules/appdelegate-subscribers.md

Reply:
Hừm... 😢 I don't known how to add RNBranch as pod depends. Look like I have to copied source from upstream [like expo-branch did](https://github.com/expo/expo/tree/master/packages/expo-branch/ios/EXBranch/RNBranch).

Reply:
assuming `react-native-branch` can be autolinked, you can just add the dependency in your podspec.

```
Pod::Spec.new do |s|
  s.dependency 'react-native-branch'
end
```

Reply:
Hi @Kudo, please release expo/expo#15550 with the tag `next`. So, I can check the config plugin locally. I also open PR in `react-native-branch` for podspec modify, see https://github.com/BranchMetrics/react-native-branch-deep-linking-attribution/pull/683

Reply:
After upgrade `apps/app` to SDKv44 includes `expo-modules-core@0.6.3`, patch `react-native-gesture-handler@2.1.0` for Android and patch `react-native-branch@5.0.0` for iOS. Now, both apps was build success. 😄 

Time for real-testing